tar/lib/src/reader.dart

```diff
@@ -0,0 +1,424 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:chunked_stream/chunked_stream.dart';
+import 'package:collection/collection.dart';
+
+import 'constants.dart';
+import 'exceptions.dart';
+import 'header.dart';
+import 'stream.dart';
+import 'utils.dart';
+
+/// [TarReader] provides sequential access to the TAR files in a TAR archive.
+/// It is designed to read from a stream and to spit out substreams for
+/// individual file contents in order to minimize the amount of memory needed
+/// to read each archive where possible.
+class TarReader {
```

Comment:
I am not entirely sure, so just brainstorming an alternative interface:
```dart

class TarEntry extends TarHeader {
  Stream<List<int>> contents;
  TarEntry._(header, contents) ...
}

Stream<TarEntry> readTarStream(Stream<List<int>> tarStream) async* {
  final r = TarReader(tarStream);
  while(r.next()) {
    yield TarEntry(r.header, r.contents);
  }
}
```

So then users would do:
```dart
await for(final entry in readTarStream(myTarStream)) {
  print(entry.name);
  await entry.drain();
}
```

Maybe it's a little crazy... because you in `TarReader` we can always make `TarReader.next()` either: drain the stream if not already drained, or throw a StateError if not drained.

Reply:
It's kinda strange, and I still need some digging to understand why, but it seems that the current implementation doesn't require us to check and drain `_contents` - somehow `ChunkedStreamIterator.substream` handles this perfectly. 

either way, I think I should be able to convert to your alternative interface pretty easily - just let me know!

Reply:
I also think that a stream of entries is slightly more "dart-y"